src/registry/artifactory.rs

```diff
@@ -25,72 +29,99 @@ pub struct Artifactory {
-        let repositories_url: Url = {
-            let mut uri = self.registry.to_owned();
-            let path = &format!("{}/api/repositories", uri.path());
-            uri.set_path(path);
-            uri.into()
-        };
-
-        let mut request = self.client.get(repositories_url);
+    async fn make_auth_request(
+        &self,
+        method: Method,
+        url: Url,
+        body: Option<Bytes>,
+    ) -> miette::Result<Response> {
+        let mut request_builder = self.client.request(method.clone(), url.clone());
 
         if let Some(token) = &self.token {
-            request = request.bearer_auth(token);
+            request_builder = request_builder.bearer_auth(token);
         }
 
-        let response = request.send().await?;
+        if let Some(body) = body {
+            request_builder = request_builder.body(body);
+        }
 
-        let status = response.status();
+        let response = request_builder.send().await.into_diagnostic()?;
 
-        if !status.is_success() {
-            tracing::info!("Artifactory response header: {:?}", response);
-        }
+        ensure!(
+            !response.status().is_redirection(),
+            "remote server attempted to redirect request - is this registry URL valid? {}",
+            self.registry
+        );
+
+        ensure!(
+            response.status() != 401,
+            "unauthorized - please provide registry credentials with `buffrs login`"
+        );
 
-        ensure!(status.is_success(), "Failed to ping artifactory");
+        response.error_for_status().into_diagnostic()
+    }
 
-        tracing::debug!("Pinging artifactory succeeded");
+    /// Pings artifactory to ensure registry access is working
+    pub async fn ping(&self) -> miette::Result<()> {
+        let repositories_url: Url = {
+            let mut uri = self.registry.to_owned();
+            let path = &format!("{}/api/repositories", uri.path());
+            uri.set_path(path);
+            uri.into()
+        };
 
-        Ok(())
+        self.make_auth_request(Method::GET, repositories_url, None)
+            .await
+            .map(|_| ())
+            .map_err(miette::Report::from)
     }
 
     /// Downloads a package from artifactory
-    pub async fn download(&self, dependency: Dependency) -> eyre::Result<Package> {
+    pub async fn download(&self, dependency: Dependency) -> miette::Result<Package> {
         ensure!(
             dependency.manifest.version.comparators.len() == 1,
-            "{} uses unsupported semver comparators",
-            dependency.package
+            UnsupportedVersionRequirement(dependency.manifest.version)
         );
 
         let version = dependency
             .manifest
             .version
             .comparators
             .first()
-            .wrap_err("internal error")?;
+            // validated above
+            .expect("unexpected error: empty comparators vector in VersionReq");
```

Comment:
I have to say this buggs me a lot, people will make errors and then we will have a runtime panic.

Reply:
Can we please please not panic? 🙇‍♀️

Reply:
Refactor done.

Reply:
From experience, someone will remove the check, without removing the panic. I think its not controversial to say, never panic in a library or application below the main function